java/piranha/src/test/java/com/uber/piranha/EnumConstantTest.java

```diff
@@ -33,6 +33,464 @@
+        BugCheckerRefactoringTestHelper.newInstance(new XPFlagCleaner(b.build()), getClass());
+
+    bcr = bcr.setArgs("-d", temporaryFolder.getRoot().getAbsolutePath());
+
+    bcr.addInputLines(
+            "TestExperimentName.java",
+            "package com.uber.piranha;",
+            "public enum TestExperimentName {",
+            " OTHER_FLAG(\"other\"),",
+            " STALE_FLAG(\"stale.flag\");",
+            " private final String key;",
+            " public String getKey() {",
+            "   return key;",
+            " }",
+            " TestExperimentName(final String key) {",
+            "  this.key = key;",
+            " }",
+            "}")
+        .addOutputLines(
+            "TestExperimentName.java",
+            "package com.uber.piranha;",
+            "public enum TestExperimentName {",
+            " OTHER_FLAG(\"other\");",
+            " private final String key;",
+            " public String getKey() {",
+            "   return key;",
+            " }",
+            " TestExperimentName(final String key) {",
+            "  this.key = key;",
+            " }",
+            "}")
+        .doTest();
+  }
+
+  /**
+   * Test for when the last enum constant is removed, and there are comments between enum constants.
+   *
+   * <p>Piranha doesn't know if it can associate the prior comment with the stale flag, so to be
+   * safe, we leave it.
+   *
+   * <p>See <a href="https://github.com/uber/piranha/issues/143">Issue #143</a>
+   */
+  @Test
+  public void cleanUpLastEnumTestWithSemicolonWithComments() {
+    ErrorProneFlags.Builder b = ErrorProneFlags.builder();
+    b.putFlag("Piranha:FlagName", "stale.flag");
+    b.putFlag("Piranha:IsTreated", "false");
+    b.putFlag("Piranha:Config", "config/properties.json");
+
+    BugCheckerRefactoringTestHelper bcr =
+        BugCheckerRefactoringTestHelper.newInstance(new XPFlagCleaner(b.build()), getClass());
+
+    bcr = bcr.setArgs("-d", temporaryFolder.getRoot().getAbsolutePath());
+
+    bcr.addInputLines(
+            "TestExperimentName.java",
+            "package com.uber.piranha;",
+            "public enum TestExperimentName {",
+            " // This is another flag!",
+            " OTHER_FLAG(\"other\"),",
+            " // This flag is stale!",
+            " STALE_FLAG(\"stale.flag\");",
+            " private final String key;",
+            " public String getKey() {",
+            "   return key;",
+            " }",
+            " TestExperimentName(final String key) {",
+            "  this.key = key;",
+            " }",
+            "}")
+        .addOutputLines(
+            "TestExperimentName.java",
+            "package com.uber.piranha;",
+            "public enum TestExperimentName {",
+            " // This is another flag!",
+            " OTHER_FLAG(\"other\");",
+            // Piranha doesn't know if it can associate this comment with the stale flag, so to be
+            // safe, we leave it
+            " // This flag is stale!",
```

Comment:
OK yeah this is something which may be a good idea to add upstream to error prone because it looks like their `replaceIncludingComments(TreePath path, String replacement, VisitorState state)` doesn't allow for end position adjustment (it probably makes sense to not include start position adjustments though!) :)

Alternatively, we find the place in the tree traversal where the enum constant TreePath statement includes the trailing semicolon so that we won't need the adjustment, if that's something that's possible. So I'll create an issue that can be followed up on.

Reply:
Made issue ticket: https://github.com/uber/piranha/issues/148

Reply:
If we do the rewriting at the class/enum level, rather than the member levels, we could always say something like, for:

```
            " // This is another flag!",
            " OTHER_FLAG(\"other\"),",
            " // This flag is stale!",
            " STALE_FLAG(\"stale.flag\");",
            " private final String key;",
```

"Remove from 1 past the end position of `OTHER_FLAG(\"other\")` to the end position of `STALE_FLAG(\"stale.flag\")`" and we would end with the natural result of:

```
            " // This is another flag!",
            " OTHER_FLAG(\"other\");",
            " private final String key;",
```

Between that and empty enum deletion/marking (see below), I am starting to think that refactoring this to work on the full enum at a time might be a good idea...

That said, It is definitely dangerous, though, since rewriting at the `enum`/`class` level can mean - depending on how the rewriting is done - that further rewrites inside its members (e.g. methods) are impossible. Maybe the right approach is still to combine what you are doing in this PR with:
a) Some way of adjusting the initial position of the deletion to "look-back" through any comment lines directly above the enum (possibly a follow up PR?)
b) Deletion of enums with no remaining constants (and _this_ can be handled at the enum level, similar to the treatment groups enums example below)